contracts/debtAllocators/YieldManager/YieldManager.sol

```diff
@@ -0,0 +1,388 @@
+     *   - Be ordered so that all debt decreases are at the beginning of the array
+     *       and debt increases at the end.
+     *
+     * This will not do any APR checks and assumes the sender has completed
+     * any and all necessary checks before sending.
+     *
+     * @param _vault The address of the vault to propose an allocation for.
+     * @param _newAllocations Array of strategies and their new proposed allocation.
+     */
+    function updateAllocationPermissioned(
+        address _vault,
+        Allocation[] memory _newAllocations
+    ) public virtual onlyGovernance {
+        // Move funds
+        _allocate(_vault, _newAllocations);
+    }
+
+    /**
+     * @notice Update a `_vault`s allocation of debt.
+     * @dev This takes the address of a vault and an array of
+     * its strategies and their specific allocation.
+     *
+     * The `_newAllocations` array should:
+     *   - Contain all strategies that hold any amount of debt from the vault
+     *       even if the debt wont be adjusted in order to get the correct
+     *       on chain APR.
+     *   - Be ordered so that all debt decreases are at the beginning of the array
+     *       and debt increases at the end.
+     *
+     * It is expected that the proposer does all needed checks for values such
+     * as max_debt, maxWithdraw, min total Idle etc. that are enforced on debt
+     * updates at the vault level.
+     *
+     * @param _vault The address of the vault to propose an allocation for.
+     * @param _newAllocations Array of strategies and their new proposed allocation.
+     */
+    function updateAllocation(
+        address _vault,
+        Allocation[] memory _newAllocations
+    )
+        external
+        virtual
+        onlyProposersOrOpen
+        returns (uint256 _currentYield, uint256 _afterYield)
+    {
+        address allocator = vaultAllocator[_vault];
+        require(allocator != address(0), "vault not added");
+
+        // Get the total assets the vault has.
+        uint256 _totalAssets = IVault(_vault).totalAssets();
+
+        // If 0 nothing to do.
+        if (_totalAssets == 0) return (0, 0);
+
+        // Always first account for the amount idle in the vault.
+        uint256 _accountedFor = IVault(_vault).totalIdle();
+        // Create local variables used through loops.
+        address _strategy;
+        uint256 _currentDebt;
+        uint256 _newDebt;
+        for (uint256 i = 0; i < _newAllocations.length; ++i) {
+            _strategy = _newAllocations[i].strategy;
+            _newDebt = uint256(_newAllocations[i].newDebt);
+            // Get the debt the strategy current has.
+            _currentDebt = IVault(_vault).strategies(_strategy).current_debt;
+            // Add to what we have accounted for.
+            _accountedFor += _currentDebt;
+
+            // Get the current weighted APR the strategy is earning
+            uint256 _strategyApr = (aprOracle.getStrategyApr(_strategy, 0) *
```

Comment:
not really apr, but i'll give you a pass

Reply:
question, what units is getStrategyApr in? I guess we don't care about the units here, but if we wanted it to be in token terms, we would need to divide by the apr units.

Reply:
its denominated in 1e18 for every vault

Reply:
Updated all the comments and variables to use "rate" instead of APR or yield https://github.com/yearn/vault-periphery/pull/30/commits/fc3b236c152d06d93bc4dcffff3b8bd14d6608b4